### PR TITLE
Add register classes

### PR DESCRIPTION
**Pull request by k-sareen**
This adds register classes which can be used to discriminate against floating-point or integer registers. Have to workaround the unstable move insertion order introduced by the hashing function.

Future PRs build on the above in order to add SSE support.

~Kunal

### CODE REVIEW

File: aeneas/src/mach/MachBackend.v3 [hunk 1]
```diff
@@ -19,6 +19,33 @@ class MachMoves {
 	var valMoves: List<(Val, int)>; // value -> usepos moves, before
 	var after: MoveResolver;	// phys -> phys moves, after
 }
+// Machine-independent representation of virtual register classes
+enum RegClass(kind: byte) {
```
Comment (titzer) [inline aeneas/src/mach/MachBackend.v3:23]: Actually, now that I think about it, we can just use the implicitly-declared "tag" for the register class.
Let's call this first one "REF" as well.

File: aeneas/src/x86/X86RegSet.v3 [hunk 2]
```diff
@@ -47,38 +73,89 @@ component X86MachRegs {
...
+			match (Regs.toRegClass(paramTypes[i])) {
+				Ptr, I32 => {
+					if (iprm < paramRegs.length) ploc[i] = paramRegs[iprm++];
+					else ploc[i] = spillStart + pspill++;
+				}
+				F32 => { // TODO: make F32 also spill 2 slots
+					if (fprm < floatParamRegs.length) ploc[i] = floatParamRegs[fprm++];
+					else ploc[i] = spillStart + pspill++;
+				}
+			} else { // I64 or F64
```
Comment (titzer) [inline aeneas/src/x86/X86RegSet.v3:130]: I think it's best to spell these out explicitly in the match statement.

File: aeneas/src/x86/X86RegSet.v3 [hunk 3]
```diff
@@ -47,38 +73,89 @@ component X86MachRegs {
...
+				match (Regs.toRegClass(returnTypes[i])) {
+					Ptr, I32 => {
+						if (iprm < retRegs.length) rloc[i] = retRegs[iprm++];
+						else rloc[i] = spillStart + rspill++;
+					}
+					F32 => { // TODO: make F32 also spill 2 slots
+						if (fprm < floatRetRegs.length) rloc[i] = floatRetRegs[fprm++];
+						else rloc[i] = spillStart + rspill++;
+					}
+				} else { // I64 or F64
```
Comment (titzer) [inline aeneas/src/x86/X86RegSet.v3:152]: Same here.

**Review by titzer — CHANGES_REQUESTED** [review]

File: aeneas/src/mach/MachBackend.v3 (hunk 1 above)
Comment (k-sareen) [reply aeneas/src/mach/MachBackend.v3:23]: You mean the enum tag? Are you saying we should replace the RegClass field in VReg with an int that represents the tag?

File: aeneas/src/x86/X86RegSet.v3 (hunk 3 above)
Comment (k-sareen) [reply aeneas/src/x86/X86RegSet.v3:152]: Done

File: aeneas/src/x86/X86RegSet.v3 (hunk 2 above)
Comment (k-sareen) [reply aeneas/src/x86/X86RegSet.v3:130]: Done

File: aeneas/src/mach/MachBackend.v3 (hunk 1 above)
Comment (k-sareen) [reply aeneas/src/mach/MachBackend.v3]: Done.

File: aeneas/src/arm/ArmRegSet.v3 [hunk 4]
```diff
@@ -65,7 +65,10 @@ component ArmMachRegs {
 		rnames[GPR] = "{gpr}";
 		rnames[ALL] = "{all}";
 
-		regs = MachRegSet.new(12, rarray, rnames, R12, CALLER_SPILL_START, CALLEE_SPILL_START);
+		var rclasses = Array<byte>.new(1);
```
Comment (titzer) [inline aeneas/src/arm/ArmRegSet.v3:68]: You should be able to just do this in one line with [GPR].

File: aeneas/src/mach/MachBackend.v3 [hunk 5]
```diff
@@ -19,6 +19,33 @@ class MachMoves {
 	var valMoves: List<(Val, int)>; // value -> usepos moves, before
 	var after: MoveResolver;	// phys -> phys moves, after
 }
+// Machine-independent representation of virtual register classes
+enum RegClass() {
```
Comment (titzer) [inline aeneas/src/mach/MachBackend.v3:23]: No need for parentheses here.

**Review by titzer — COMMENTED** [review]
lgtm with minor formatting changes

File: aeneas/src/mach/MachBackend.v3 (hunk 5 above)
Comment (k-sareen) [reply aeneas/src/mach/MachBackend.v3:23]: Ah missed that somehow. Done.

File: aeneas/src/arm/ArmRegSet.v3 (hunk 4 above)
Comment (k-sareen) [reply aeneas/src/arm/ArmRegSet.v3]: Done

**Pull request merged**